scanpy/tests/test_metrics.py

```diff
@@ -0,0 +1,97 @@
+from operator import eq
```

Comment:
Will this still let pytest show diffs when tests fail as if you had used `==`?

Reply:
Yes. The diffs are a bit different, but it still shows the values.

Reply:
Actually very different and much less helpful:

```py
from operator import eq

def test_op():
	assert [0,2,2] == [0,1,2]


def test_fn():
	assert eq([0,2,2], [0,1,2])
```

``pytb
================== test session starts ===================
platform linux -- Python 3.8.0, pytest-5.2.4, py-1.8.0, pluggy-0.13.1
rootdir: ~/Dev/Python
plugins: cov-2.8.1, black-0.3.7, hypothesis-4.46.1, flake8-1.0.4
collected 2 items                                        

test_eq_vs_operator.py FF                          [100%]

======================== FAILURES ========================
________________________ test_op _________________________

    def test_op():
>       assert [0,2,2] == [0,1,2]
E    assert [0, 2, 2] == [0, 1, 2]
E      At index 1 diff: 2 != 1
E      Use -v to get the full diff

test_eq_vs_operator.py:4: AssertionError
________________________ test_dn _________________________

    def test_fn():
>       assert eq([0,2,2], [0,1,2])
E    assert False
E     +  where False = eq([0, 2, 2], [0, 1, 2])

test_eq_vs_operator.py:8: AssertionError
=================== 2 failed in 0.04s ====================
```

Reply:
Each time I use `eq` here, it's comparing scalars. I mainly used it for the formatting here, so I'm easy on changing it, but I don't think it'll give different results.

I would definitely support `eq` giving the same results as `==` in general, which could be added to `pytest`.

Reply:
Scanpydoc throws an error if I try to add this to the docs. The full log is below, but I think these are the relevant parts of the traceback:

```python
  File "/usr/local/lib/python3.7/site-packages/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/usr/local/lib/python3.7/site-packages/scanpydoc/elegant_typehints/return_tuple.py", line 46, in process_docstring
    ret_types = get_tuple_annot(get_type_hints(obj).get("return"))
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/typing.py", line 996, in get_type_hints
    'or function.'.format(obj))
TypeError: <dispatched gearys_c> is not a module, class, method, or function.
```

<details>
<summary> Full error </summary>

```python
# Sphinx version: 2.2.1
# Python version: 3.7.5 (CPython)
# Docutils version: 0.15.2 release
# Jinja2 version: 2.10.1
# Last messages:
#   loading pickled environment...
#   done
#   [autosummary] generating autosummary for: api/index.rst, api/plotting.rst, api/scanpy.Neighbors.compute_eigen.rst, api/scanpy.Neighbors.compute_neighbors.rst, api/scanpy.Neighbors.compute_transitions.rst, api/scanpy.Neighbors.connectivities.rst, api/scanpy.Neighbors.distances.rst, api/scanpy.Neighbors.distances_dpt.rst, api/scanpy.Neighbors.eigen_basis.rst, api/scanpy.Neighbors.eigen_values.rst, ..., external/scanpy.external.tl.cyclone.rst, external/scanpy.external.tl.palantir.rst, external/scanpy.external.tl.phate.rst, external/scanpy.external.tl.phenograph.rst, external/scanpy.external.tl.sandbag.rst, index.rst, installation.rst, references.rst, release_notes.rst, tutorials.rst
#   [autosummary] generating autosummary for: /Users/isaac/github/scanpy/docs/api/scanpy.metrics.confusion_matrix.rst, /Users/isaac/github/scanpy/docs/api/scanpy.metrics.gearys_c.rst
#   building [mo]: targets for 0 po files that are out of date
#   building [html]: targets for 6 source files that are out of date
#   updating environment:
#   [extensions changed ('5')]
#   242 added, 0 changed, 0 removed
#   reading sources... [  0%] api/index
# Loaded extensions:
#   sphinx.ext.mathjax (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/mathjax.py
#   sphinxcontrib.applehelp (1.0.1) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/applehelp/__init__.py
#   sphinxcontrib.devhelp (1.0.1) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/devhelp/__init__.py
#   sphinxcontrib.htmlhelp (1.0.2) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/htmlhelp/__init__.py
#   sphinxcontrib.serializinghtml (1.1.3) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/serializinghtml/__init__.py
#   sphinxcontrib.qthelp (1.0.2) from /usr/local/lib/python3.7/site-packages/sphinxcontrib/qthelp/__init__.py
#   alabaster (0.7.12) from /usr/local/lib/python3.7/site-packages/alabaster/__init__.py
#   sphinx.ext.autodoc (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py
#   sphinx.ext.intersphinx (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/intersphinx.py
#   sphinx.ext.doctest (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/doctest.py
#   sphinx.ext.coverage (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/coverage.py
#   sphinx.ext.napoleon (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/napoleon/__init__.py
#   sphinx.ext.autosummary (2.2.1) from /usr/local/lib/python3.7/site-packages/sphinx/ext/autosummary/__init__.py
#   sphinx_autodoc_typehints (unknown version) from /usr/local/lib/python3.7/site-packages/sphinx_autodoc_typehints.py
#   scanpydoc.autosummary_generate_imported (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/autosummary_generate_imported.py
#   scanpydoc.definition_list_typed_field (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/definition_list_typed_field.py
#   scanpydoc.elegant_typehints (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/elegant_typehints/__init__.py
#   scanpydoc.rtd_github_links (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/rtd_github_links.py
#   scanpydoc (0.4.4) from /usr/local/lib/python3.7/site-packages/scanpydoc/__init__.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/sphinx/cmd/build.py", line 276, in build_main
    app.build(args.force_all, filenames)
  File "/usr/local/lib/python3.7/site-packages/sphinx/application.py", line 346, in build
    self.builder.build_update()
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 299, in build_update
    len(to_build))
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 311, in build
    updated_docnames = set(self.read())
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 418, in read
    self._read_serial(docnames)
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 439, in _read_serial
    self.read_doc(docname)
  File "/usr/local/lib/python3.7/site-packages/sphinx/builders/__init__.py", line 479, in read_doc
    doctree = read_doc(self.app, self.env, self.env.doc2path(docname))
  File "/usr/local/lib/python3.7/site-packages/sphinx/io.py", line 326, in read_doc
    pub.publish()
  File "/usr/local/lib/python3.7/site-packages/docutils/core.py", line 217, in publish
    self.settings)
  File "/usr/local/lib/python3.7/site-packages/sphinx/io.py", line 114, in read
    self.parse()
  File "/usr/local/lib/python3.7/site-packages/docutils/readers/__init__.py", line 77, in parse
    self.parser.parse(self.input, document)
  File "/usr/local/lib/python3.7/site-packages/sphinx/parsers.py", line 95, in parse
    self.statemachine.run(inputlines, document, inliner=self.inliner)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 171, in run
    input_source=document['source'])
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2771, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 395, in new_subsection
    node=section_node, match_titles=True)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2771, in underline
    self.section(title, source, style, lineno - 1, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 327, in section
    self.new_subsection(title, lineno, messages)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 395, in new_subsection
    node=section_node, match_titles=True)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 282, in nested_parse
    node=node, match_titles=match_titles)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 196, in run
    results = StateMachineWS.run(self, input_lines, input_offset)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 239, in run
    context, state, transitions)
  File "/usr/local/lib/python3.7/site-packages/docutils/statemachine.py", line 460, in check_line
    return method(match, context, next_state)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2344, in explicit_markup
    nodelist, blank_finish = self.explicit_construct(match)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2356, in explicit_construct
    return method(self, expmatch)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2099, in directive
    directive_class, match, type_name, option_presets)
  File "/usr/local/lib/python3.7/site-packages/docutils/parsers/rst/states.py", line 2148, in run_directive
    result = directive_instance.run()
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autosummary/__init__.py", line 240, in run
    items = self.get_items(names)
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autosummary/__init__.py", line 349, in get_items
    documenter.add_content(None)
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py", line 471, in add_content
    for i, line in enumerate(self.process_doc(docstrings)):
  File "/usr/local/lib/python3.7/site-packages/sphinx/ext/autodoc/__init__.py", line 441, in process_doc
    self.options, docstringlines)
  File "/usr/local/lib/python3.7/site-packages/sphinx/application.py", line 450, in emit
    return self.events.emit(event, *args)
  File "/usr/local/lib/python3.7/site-packages/sphinx/events.py", line 103, in emit
    results.append(callback(self.app, *args))
  File "/usr/local/lib/python3.7/site-packages/scanpydoc/elegant_typehints/return_tuple.py", line 46, in process_docstring
    ret_types = get_tuple_annot(get_type_hints(obj).get("return"))
  File "/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7/typing.py", line 996, in get_type_hints
    'or function.'.format(obj))
TypeError: <dispatched gearys_c> is not a module, class, method, or function.

```

</details>

Reply:
OK, if you use it on scalars it’s fine, but keep in mind that for 1D sequences, nothing should beat `==`.

Reply:
Well, unless they are broadcast able. It would also be nice if np.allclose would return that kind of output too.